<a href="https://colab.research.google.com/github/AnjulaMehto/UCC304_AI/blob/main/Learn_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
# **Learn Python **
---

### 16 programs to explain the various concepts in python programming such as:

- Syntax,
- Loop,
- if-else,
- Data Structures,
- Strings,
- File Handling,
- Exception Handling,
- Random Numbers,
- Command Line Argunment
- Use of Libraries

---
### **Self learning resource**
Tutorial on Python (Byte of Python) **<a href="https://python.swaroopch.com/"> Click Here</a>**


```



```




---
# **1 Hello World**
---
**Learning:** How to print and run python program

In [1]:
print ("Hello, Python!")


Hello, Python!


---
# **2 Add numbers and Concatinate strings**
**Learning:** How to declare variable, add, concatenate and print the result.

---
### **2.1 Add two numbers**


In [2]:
a = 15
b = 340
c = a + b       # Add two numbers
print (a, " + ", b, " --> ", c)


15  +  340  -->  355



---
### **2.2 Concatinate two strings**
---

In [3]:
a = "Chandra"
b = " Shekhar"
c = a + b		# Concatinate two strings
print (a, " + ", b, " --> ", c)


Chandra  +   Shekhar  -->  Chandra Shekhar



---
### **2.3 Concatinate string with number**
---

In [4]:
a = "Chandra"
b = 250
c = a + str(b)		# Concatinate string with number
print (a, " + ", b, " --> ", c)


Chandra  +  250  -->  Chandra250


---
# **3 Input from user**
**Learning:** How to take input from user

---
### **3.1 Input two strings from user and concatinate them**
---

In [5]:
a = input("Enter First String: ")
b = input("Enter Second String: ")
c = a + b		# concatinate two strings
print  (a, " + ", b, "      --> ", c)

# Run the program with (1) Two strings and (2) Two numbers


keshav  +  goyal       -->  keshavgoyal


---
### **3.2 Input two numbers from user and add them**
---

In [6]:
a = int(input("Enter First No: "))
b = int(input("Enter Second No: "))
c = a + b
print (a, " + ", b, " --> ", c)


5  +  7  -->  12


---
# **4 Loop**
**Learning:** Learn various loops.

---
### **4.1 While Loop**

In [7]:
i=1
while i <= 15:
	print (i)
	i = i+1


1
2
3
4
5
6
7
8
9
10
11
12
13
14
15


---
### **4.2 Range Function**
---

In [8]:
print ("range(12)          --> ", list(range(12)))
print ("range(5,15)        --> ", list(range(5,15)))
print ("range(3,21,3)      --> ", list(range(3,21,3)))
print ("range(-5,-25,3)    --> ", list(range(-5,-25,3)))
print ("range(-5,-25,-3)   --> ", list(range(-5,-25,-3)))


range(12)          -->  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
range(5,15)        -->  [5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
range(3,21,3)      -->  [3, 6, 9, 12, 15, 18]
range(-5,-25,3)    -->  []
range(-5,-25,-3)   -->  [-5, -8, -11, -14, -17, -20, -23]


---
### **4.3 For loop**
---
### **4.3.1 For loop - Version 1**

In [9]:
for i in range(0,15):
	print (i)


0
1
2
3
4
5
6
7
8
9
10
11
12
13
14


---
### **4.3.2 For loop - Version 2**
---

In [10]:
for i in range(0,30,3):
	print (i)


0
3
6
9
12
15
18
21
24
27


---
### **4.3.3 For loop - Version 3**
---

In [11]:
for i in range(0,-15,-1):
	print (i)


0
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14


---
### **4.4 Print table of 5**
---

In [12]:
for i in range(1,11):
	print (8," * ", i , " = ", i * 8)


8  *  1  =  8
8  *  2  =  16
8  *  3  =  24
8  *  4  =  32
8  *  5  =  40
8  *  6  =  48
8  *  7  =  56
8  *  8  =  64
8  *  9  =  72
8  *  10  =  80


---
### **4.5 Sum all numbers from 1 to 10**
---
### **4.5.1 Version 1**

In [13]:
s=0
for i in range(1,21):
	s=s+i
print ("Sum is --> ",s)


Sum is -->  210


---
### **4.5.2 Version 2**
---

In [14]:
print ("Sum is --> ", sum(range(1,21)))


Sum is -->  210


---
# **5 If-Else - Conditional Checking**
**Learning:** if-else Condition

---
### **5.1 Input two numbers from user and compare them**

In [15]:
a = int(input("Enter First No: "))
b = int(input("Enter Second No: "))
if  a > b:
	print (a," > ",b)
else:
	print (a," < ",b)


10  <  20


---
### **5.2 Check weather a number is odd or even**
---

In [16]:
n = int(input("Enter a No: "))
if  n % 2 == 0:
	print (n," is even")
else:
	print (n," is odd")


9  is odd


---
### **5.3 Check weather a number is prime of not**
---

In [17]:
n = int(input("Enter a No: "))
f=0
for i in range(2, n//2 + 1):
	if  n % i == 0:
		f=1
		break

if f==0:
	print ("Prime")
else:
	print ("Not Prime")


Prime


---
### **5.4 Conditional Checking - Compare strings**
---

In [18]:
a = input("Enter First String : ")
b = input("Enter Second String: ")

if a == b:
	print ("a == b")
elif a >= b:
	print ("a > b")
else:
	print ("a < b")


a < b


---
# **6 Functions**
**Learning:** How to declare and call function

---
### **6.1 Add two numbers**

In [19]:
def Add(a,b):
	c=a+b
	return c

print ("Add(15,25) -->", Add(15,25))
print ("Add(30,60) -->", Add(30,60))
print ("Add(90,300) -->", Add(90,300))


Add(15,25) --> 40
Add(30,60) --> 90
Add(90,300) --> 390


---
### **6.2 Prime number**
---

In [20]:
def IsPrime(n):
	for i in range(2, n//2 + 1):
		if n%i==0:
			return 0
	return 1

print ("IsPrime(18)  --> ", IsPrime(18))
print ("IsPrime(29)  --> ", IsPrime(29))
print ("IsPrime(150) --> ", IsPrime(150))
print ("IsPrime(41)  --> ", IsPrime(41))


IsPrime(18)  -->  0
IsPrime(29)  -->  1
IsPrime(150) -->  0
IsPrime(41)  -->  1


---
### **6.3 Add 1 to n**
---

In [21]:
def AddN(n):
	s= sum(range(n+1))
	return s

print ("AddN(15)  --> ", AddN(15))
print ("AddN(25)  --> ", AddN(25))
print ("AddN(60)  --> ", AddN(60))
print ("AddN(150) --> ", AddN(150))


AddN(15)  -->  120
AddN(25)  -->  325
AddN(60)  -->  1830
AddN(150) -->  11325


---
# **7 Math library**
**Learning:** Use math library

---

In [22]:
import math as m
print ("exp(-150)    --> ", m.exp(-150))  # Exponential function
print ("log(200,2)   --> ", m.log(200,2)) # Log
print ("log(200,10)  --> ", m.log(200,10))# Log
print ("log10(200)   --> ", m.log10(200)) # Log 10
print ("m.cos(45)    --> ", m.cos(45))    # cos
print ("m.sin(45)    --> ", m.sin(45))    # sin
print ("m.tan(45)    --> ", m.tan(45))    # tan
print ("m.sqrt(441)  --> ", m.sqrt(441))
print ("m.ceil(72.3) --> ", m.ceil(72.3))
print ("m.floor(72.3)--> ", m.floor(72.3))


exp(-150)    -->  7.175095973164411e-66
log(200,2)   -->  7.643856189774724
log(200,10)  -->  2.301029995663981
log10(200)   -->  2.3010299956639813
m.cos(45)    -->  0.5253219888177297
m.sin(45)    -->  0.8509035245341184
m.tan(45)    -->  1.6197751905438615
m.sqrt(441)  -->  21.0
m.ceil(72.3) -->  73
m.floor(72.3)-->  72


---
# **8 Strings**
**Learning:** How to handle string

---
### **8.1 Indexing in string**

In [23]:
var = 'Good Morning!'
print ("var      --> ", var)
print ("var[0]   --> ", var[0])
print ("var[1:5] --> ", var[1:5])
print ("var[:-5] --> ", var[:-5])


var      -->  Good Morning!
var[0]   -->  G
var[1:5] -->  ood 
var[:-5] -->  Good Mor


---
### **8.2 String length, upper, lower**
---

In [24]:
var = 'Good Morning!'
print ("String --> ", var)
print ("Length --> : ", len(var))
print ("Upper  --> : ", var.upper())
print ("Lower  --> : ", var.lower())


String -->  Good Morning!
Length --> :  13
Upper  --> :  GOOD MORNING!
Lower  --> :  good morning!


---
### **8.3 String formatting**
---

In [25]:
name=input("Enter your name: ")
age=int(input("Enter your age : "))
price=float(input("Enter the book price: "))
s="\nYour name is %s, age is %d and book price is %f" %(name.upper(),age,price)
print (s)



Your name is KESHAV GOYAL, age is 18 and book price is 50000.000000


---
### **8.4 String in Triple Quotes**
---

In [26]:
para_str = """This is a long string that is made up of
several lines and non-printable characters such as
TAB ( \t ) and they will show up that way when displayed.
NEWLINEs within the string, whether explicitly given like
this within the brackets [ \n ], or just a NEWLINE within
the variable assignment will also show up.
"""
print (para_str)

This is a long string that is made up of
several lines and non-printable characters such as
TAB ( 	 ) and they will show up that way when displayed.
NEWLINEs within the string, whether explicitly given like
this within the brackets [ 
 ], or just a NEWLINE within
the variable assignment will also show up.



---
### **8.5 String strip**
---

In [27]:
var =" Global   Village    "

print("String    --> ", var)
print("Length    --> ", len(var))
print("var strip --> ", var.strip())
print("Length of var after strip --> ", len(var.strip()))


String    -->   Global   Village    
Length    -->  21
var strip -->  Global   Village
Length of var after strip -->  16


---
### **8.6 String split**
---

In [28]:
var =" Global,   Village    "

print("String    --> ", var)
print("Length    --> ", len(var))
print("var split --> ", var.split())
print("var split --> ", var.split(' '))
print("var split --> ", var.split(','))

# Strip + Split
print("var split --> ", var.strip().split(','))


String    -->   Global,   Village    
Length    -->  22
var split -->  ['Global,', 'Village']
var split -->  ['', 'Global,', '', '', 'Village', '', '', '', '']
var split -->  [' Global', '   Village    ']
var split -->  ['Global', '   Village']


---
### **8.7 Count in string**
---

In [29]:
var=" Global Village    "
print ("String       --> ", var)
print ("Count of ' ' --> ", var.count(' '))
print ("Count of 'a' --> ", var.count('a'))
print ("Count of 'l' --> ", var.count('l'))


String       -->   Global Village    
Count of ' ' -->  6
Count of 'a' -->  2
Count of 'l' -->  4


---
### **8.8 Reverse a String**
---

In [30]:
var="Global Village"
print ("String    --> ", var)
print ("var[::1]  --> ", var[::1])
print ("var[::2]  --> ", var[::2])
print ("var[::-1] --> ", var[::-1])
print ("var[::-2] --> ", var[::-2])

var=var[::-1]
print ("var after reverse --> ", var)


String    -->  Global Village
var[::1]  -->  Global Village
var[::2]  -->  Goa ilg
var[::-1] -->  egalliV labolG
var[::-2] -->  ealVlbl
var after reverse -->  egalliV labolG


---
### **8.9 Palindrome**
---

In [31]:
s1="Global Village"
s2="racecar"
s3="level"
s4="python"
print ("s1 --> ", s1==s1[::-1])
print ("s2 --> ", s2==s2[::-1])
print ("s3 --> ", s3==s3[::-1])
print ("s4 --> ", s4==s4[::-1])


s1 -->  False
s2 -->  True
s3 -->  True
s4 -->  False


---
# **9 Random Numbers/String**
**Learning:** Generate Random Numbers/String

---
### **9.1 Generate random number between 0 and 1**

In [32]:
import random as r
print (r.random())
print (r.random())
print (round(r.random(),3))


0.17632523358641727
0.8916835903957788
0.267


---
### **9.2 Generate random integer number**
---

In [33]:
import random as r
print (r.randint(1, 50))
print (r.randint(1, 50))
print (r.randint(-20, 20))
print (r.randint(-20, 20))


19
13
20
-10


---
### **9.3 Generate random real number**
---

In [34]:
import random as r
print (r.uniform(1, 50))
print (r.uniform(1, 50))
print (r.uniform (-20, 20))
print (r.uniform (-20, 20))
print (round(r.uniform (-20, 20),3))


29.56354449095786
46.669395144130156
13.325661074251734
11.80809211783047
-15.296


---
### **9.4 Select sample from a list of elements**
---

In [35]:
import random as r

A=[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

print (r.sample(A, 5))
print (r.sample(A, 2))
print (r.sample(range(0,200), 3))
print (r.sample(range(-50,50), 4))


[60, 30, 10, 50, 70]
[100, 10]
[37, 118, 27]
[-12, 37, -34, 29]


---
### **9.5 Generate random string**
---

In [36]:
import string as s
import random as r
print ("String        --> ",s.ascii_letters)

passwd=r.sample(s.ascii_letters, 8)
print ("Selected Char --> ",passwd)

passwd1="".join(passwd)
print ("passwd1       --> ",passwd1)

passwd2="+".join(passwd)
print ("passwd2       --> ",passwd2)

passwd3="*".join(passwd)
print ("passwd3       --> ",passwd3)


String        -->  abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
Selected Char -->  ['P', 'y', 'Q', 'x', 'J', 'I', 'g', 'a']
passwd1       -->  PyQxJIga
passwd2       -->  P+y+Q+x+J+I+g+a
passwd3       -->  P*y*Q*x*J*I*g*a


---
### **9.6 Generate random digits**
---

In [37]:
import string as s
import random as r
print ("Digits --> ",s.digits)

otp=r.sample(s.digits, 6)
print ("Selected num1 --> ",otp)
otp="".join(otp)
print ("otp1          --> ",otp)

otp=r.sample(s.digits, 6)
print ("Selected num2 --> ",otp)
otp="".join(otp)
print ("otp2          --> ",otp)

otp=r.sample(s.digits, 6)
print ("Selected num3 --> ",otp)
otp="".join(otp)
print ("otp3          --> ",otp)


Digits -->  0123456789
Selected num1 -->  ['3', '7', '6', '9', '0', '1']
otp1          -->  376901
Selected num2 -->  ['2', '3', '0', '5', '8', '6']
otp2          -->  230586
Selected num3 -->  ['8', '6', '1', '3', '9', '7']
otp3          -->  861397


---
### **9.7 Generate random string + digits**
---

In [38]:
import string as s
import random as r
print ("String + Digits --> ",s.ascii_letters + s.digits)

mixPasswd=r.sample(s.ascii_letters + s.digits, 6)
print ("\nSelected Str1 --> ",mixPasswd)
mixPasswd="".join(mixPasswd)
print ("mixPasswd1    --> ",mixPasswd)

mixPasswd=r.sample(s.ascii_letters + s.digits, 7)
print ("\nSelected Str2 --> ",mixPasswd)
mixPasswd="".join(mixPasswd)
print ("mixPasswd2    --> ",mixPasswd)

splChar="#@!~%^&*()_+=-[]{}|"
mixPasswd=r.sample(splChar + s.ascii_letters + s.digits, 10)
print ("\nSelected Str3 --> ",mixPasswd)
mixPasswd="".join(mixPasswd)
print ("mixPasswd3    --> ",mixPasswd)


String + Digits -->  abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789

Selected Str1 -->  ['8', 'g', 'o', 'd', 'l', 'm']
mixPasswd1    -->  8godlm

Selected Str2 -->  ['V', '8', 'x', 'h', 'q', '3', 'm']
mixPasswd2    -->  V8xhq3m

Selected Str3 -->  ['u', 'U', 'n', 'h', '*', '%', '{', 'A', 'c', 'E']
mixPasswd3    -->  uUnh*%{AcE


---
# **10 Exception Handaling**
**Learning:** How to handle exceptions

---
### **10.1 Error Generation**

In [39]:
for i in range(-6,7):
	print ("100/",i," --> ", 100/i)


100/ -6  -->  -16.666666666666668
100/ -5  -->  -20.0
100/ -4  -->  -25.0
100/ -3  -->  -33.333333333333336
100/ -2  -->  -50.0
100/ -1  -->  -100.0


ZeroDivisionError: division by zero

---
### **10.2 Exception handaling for division by zero**
---

In [ ]:
for i in range(-6,7):
	try:
		print ("100/",i," --> ", 100/i)
	except:
		print ("error")


---
### **10.3 Exception handaling for array out of index**
---

In [ ]:
L=[10,20,30,40,50,60]

for i in range(9):
	try:
		print (i," --> ",L[i])
	except:
		print ("error")


---
### **10.4 Exception handaling for file not found**
---

In [ ]:
fileName=input("Enter File Name: ")
fp=open(fileName)	# Open the file in reading mode
fp.close()
print ("Done")


---
### **10.5 Exception handaling for file not found**
---

In [ ]:
fileName=input("Enter File Name: ")
try:
	fp=open(fileName)	# Open the file in reading mode
	fp.close()
except:
	print ("Error !! \"%s\" File Not Found"%(fileName))

print ("Done")


---
# **11 Data Structure 1 - List**
**Learning:** How to use list, add, delete and search in the list.

**Note:** Read more about list and try yourself

---
### **11.1 List Declaration**


In [ ]:
L = ["Aarav",  'Mehta', 2.72,  7 ]
print ("Original List: ", L)
print ("Number of elements in list: ", len(L))


---
### **11.2 List Iteration**
---

In [ ]:
L = ["Aarav",  'Mehta', 2.72,  7 ]
print ("Original List: ", L)
i=0
while i < len(L):
	print (L[i])
	i+=1


---
### **11.3 List Iteration using for loop**
---

In [ ]:
L = ["Aarav",  'Mehta', 2.72,  7 ]
print ("Original List: ", L)
for i in range(0, len(L)):
	print (L[i])


---
### **11.4 List Iteration using for loop**
---

In [ ]:
L = ["Aarav",  'Mehta', 2.72,  7 ]
print ("Original List --> ", L)
for s in L:
	print (s)


---
### **11.5 Adding and deleting from list**
---

In [ ]:
L = ["Aarav",  'Mehta', 2.72,  7 ]
print ("Original List       --> ", L)

L.append("Kabir")
print ("List After Adding   --> ", L)

del L[1]
print ("List After Deleting --> ", L)


---
### **11.6 Sum/Average of List**
---

In [ ]:
L=[4, 8, 12, 16, 7, 4, 3]
print ("Original List --> ", L)

print ("Sum     --> ", sum(L))
print ("Average --> ", sum(L)/len(L))
print ("Average --> ", sum(L)//len(L))

print ("L * 3   --> ", L * 3)     # Every element get tripled
print ("L + L   --> ", L + L)     # Every element get doubled


---
### **11.7 Min/Max/Sort the list**
---

In [ ]:
L=[4, 8, 12, 16, 7, 4, 3]
print ("Original List --> ", L)

print ("max --> ", max(L))
print ("min --> ", min(L))

print ("\nBefore Sort            --> ", L)
L.sort()

print ("After Sort (Asending)  --> ", L)

L.sort(reverse=True)
print ("After Sort (Desending) --> ", L)


---
### **11.8 Merge lists and select elements**
---

In [ ]:
L1 = [5, 10, 15]
L2 = [20, 8, 5, 3]
L3 = L1 + L2
print ("L1 --> ",L1)
print ("L2 --> ",L2)
print ("L3 --> ",L3)

print ("\nL3[2:]  --> ",L3[2:])
print ("L3[2:5] --> ",L3[2:5])
print ("L3[:-1] --> ",L3[:-1])
print ("L3[::2] --> ",L3[::2])


---
### **11.9 Multiply all elements of the list by a constant**
---

In [ ]:
L = [20, 8, 5, 3, 10]
print ("Original List --> ", L)

newL = [ i * 4 for i in L ]
print ("After Multiply with constant --> ", newL)


---
### **11.10 Searching in the list**
---

In [ ]:
L=[5, 10, 15, 20, 8, 5, 4]
print ("Original List --> ", 10 in L)
print ("Original List --> ", 7 in L)
print ("Original List --> ", 15 in L)

if (10 in L) == True:
	print ("Present")
else:
	print ("Not Present")

if 7 in L == False:
	print ("Not Present")
else:
	print ("Present")


---
# **12 Data Structure 2 - Dictionary**
**Learning:** How to use Dictionary, add, delete, search in Dictionary

**Note:** Read more about Dictionary and try yourself

---
### **12.1 Declare Dictionary**

In [ ]:
CGPA={1:7.8, 2:6.9, 3:8.2, 5:9.0, 6:7.1}
print ("Dictionary      --> ", CGPA)
print ("Num of elements --> ", len(CGPA))

print ("CGPA of 1       --> ", CGPA[1])
print ("CGPA of 3       --> ", CGPA[3])
print ("CGPA of 5       --> ", CGPA[5])
print ("CGPA of 2       --> ", CGPA[2])


---
### **12.2 Triverse dictionary**
---

In [ ]:
CGPA={1:7.8, 2:6.9, 3:8.2, 5:9.0, 6:7.1}
for k in CGPA:
	print ("CGPA of ", k, " --> ", CGPA[k])


---
### **12.3 Getting Keys and Values**
---

In [ ]:
CGPA={1:7.8, 2:6.9, 3:8.2, 5:9.0, 6:7.1}
print ("Dictionary --> ", CGPA)
print ("Keys       --> ", list(CGPA.keys()))
print ("Values     --> ", list(CGPA.values()))


---
### **12.4 Updating, Adding and Deleting from Dictionary**
---

In [ ]:
CGPA={1:7.8,2:6.9,3:8.2,5:9.0,6:7.1}
print ("Original Dictionary --> ", CGPA)

CGPA[3] = 8.5
print ("After Updating (3)  --> ", CGPA)

CGPA[4] = 7.9
print ("After Adding (4)    --> ", CGPA)

del CGPA[1]
print ("After Deleting (1)  --> ", CGPA)

CGPA.clear()
print ("After Clear         --> ", CGPA)

del CGPA
print ("After Delete        --> ", CGPA)


---
### **12.5 Checking for Key in Dictionary**
---

In [ ]:
CGPA={1:7.8, 2:6.9, 3:8.2, 5:9.0, 6:7.1}
print ("Original Dictionary --> ", CGPA)
print ("Is Key 2 Present    --> ", 2 in CGPA)
print ("Is Key 9 Present    --> ", 9 in CGPA)


---
### **12.6 More example1**
---

In [ ]:
HomeTown={"Aarav":"Mumbai", "Isha":"Jaipur", "Karan":"Chandigarh", "Meera":"Pune"}
print ("Original Dictionary --> ", HomeTown)
print ("Home Town of Aarav is --> ", HomeTown["Aarav"])
print ("Home Town of Isha is   --> ", HomeTown["Isha"])
print ("Home Town of Karan is     --> ", HomeTown["Karan"])
print ("Home Town of Meera is   --> ", HomeTown["Meera"])


---
### **12.7 More example2**
---

In [ ]:
HomeTown={"Aarav":"Mumbai", "Isha":"Jaipur", "Karan":"Chandigarh", "Meera":"Pune"}
print ("Original Dictionary --> ", HomeTown)

for d in HomeTown:
	print ("Home Town of ", d, " is  --> ", HomeTown[d])


---
# **13 Data Structure 3 - Tuple**

**Learning:** How to use Tuple, add, delete, search in Tuple

**Note:** Read more about Tuple and try yourself

---
### **13.1 Declare Tuple**

In [ ]:
# Method 1
T = ("Aarav", 'Mehta', 2.72, 7)

print ("T               -->", T)
print ("Num of elements -->", len(T))
print ("Type of Object  -->", type(T))


In [ ]:
# Method 2
T = tuple(["Aarav", 'Mehta', 2.72, 7])   # Convert list to tuple
#T = tuple(("Aarav", 'Mehta', 2.72, 7))  # Also Works

print ("T               -->", T)
print ("Num of elements -->", len(T))
print ("Type of Object  -->", type(T))


---
### **13.2 Tuple Iteration**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T -->", T)

i = 0
while i < len(T):
    print (T[i])
    i += 1


---
### **13.3 Tuple iteration using for loop**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T -->", T)

for i in range(0, len(T)):
    print (T[i])


---
### **13.4 Tuple iteration using for loop**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T -->", T)

for s in T:
    print (s)


---
### **13.5 Accessing/Selecting in Tuple**
---

In [ ]:
# Example 1:
T = (5, 10, 15, 20, 8, 5, 4)
print ("T     -->", T)

print ("T[1]  -->", T[1])
print ("T[2]  -->", T[2])
print ("T[-1] -->", T[-1])
print ("T[-2] -->", T[-2])


In [ ]:
# Example 2:
T = (5, 10, 15, 20, 8, 5, 4)
print ("T        -->", T)

print ("T[1:3]   -->", T[1:3])
print ("T[2:]    -->", T[2:])
print ("T[2:5]   -->", T[2:5])
print ("T[:2]    -->", T[:2])
print ("T[:-1]   -->", T[:-1])
print ("T[-4:-1] -->", T[-4:-1])


---
### **13.6 Sum/Average of Tuple**
---

In [ ]:
T = (5, 10, 15, 20, 8, 5, 4)
print ("T       -->", T)
print ("Sum     -->", sum(T))
print ("Average -->", sum(T)/len(T))
print ("Average -->", sum(T)//len(T))


---
### **13.7 Min/Max in Tuple**
---

In [ ]:
# Example 1
T = (5, 10, 15, 20, 8, 5, 4)                # Integer Tuple
print ("T   -->", T)
print ("Max -->", max(T))
print ("Min -->", min(T))


In [ ]:
# Example 2
T = ("Zebra", "Apple", "Mango", "Kiwi")      # String Tuple
print ("T   -->", T)
print ("Max -->", max(T))
print ("Min -->", min(T))


---
### **13.8 Merging Tuples**
---

In [ ]:
T1 = (5, 10, 15)
T2 = (20, 8, 5, 3)

print ("T1 -->", T1)
print ("T2 -->", T2)

T3 = T1 + T2
print ("T3 -->", T3)

T4 = T1 + T2 + T1 + T2
print ("T4 -->", T4)


---
### **13.9 Merging part of Tuples**
---

In [ ]:
T1 = (5, 10, 15)
T2 = (20, 8, 5, 3)

print ("T1 -->", T1)
print ("T2 -->", T2)

T3 = T1[1:2] + T2[1:3]
print ("T3 -->", T3)

T4 = T1[:-2] + T2[:-3]
print ("T4 -->", T4)


---
### **13.10 Searching in the tuple**
---

In [ ]:
T = (5, 10, 15, 20, 8, 5, 4)
print ("T       -->", T)

print ("10 in T -->", 10 in T)
print ("7  in T -->", 7 in T)
print ("20 in T -->", 20 in T)


---
### **13.11 Adding element to Tuple (Error)**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T -->", T)

T[2] = 500               # Error; 'tuple' object does not support item assignment
print ("T -->", T)

#Tuples are unchangeable. We cannot add items to it.


---
### **13.12 Adding element to Tuple - (Jugaad)**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T         -->", T)

T1 = list(T)
T1.append(6.5)
T = tuple(T1)

print ("After Add -->", T)


---
### **13.13 Inserting element in Tuple - (Jugaad)**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T            -->", T)

T1 = list(T)
T1.insert(2, "Kabir")
T = tuple(T1)

print ("After Insert -->", T)


---
### **13.14 Deleting from Tuple (Error)**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T            -->", T)

del T[1]

print ("After Delete -->", T)


---
### **13.15 Deleting from Tuple - (Jugaad)**
---

In [ ]:
T = ("Aarav", 'Mehta', 2.72, 7)
print ("T            -->", T)

T1 = list(T)
del T1[1]
T = tuple(T1)

print ("After Delete -->", T)


---
# **14 Data Structure 4 - Set**

**Learning:** How to use Set, add, delete, search in Set

**Note:** Read more about Set and try yourself

---
### **14.1 Declare Set**

In [ ]:
s = set(['P', 'Q', 'R', 'S','R', 'S' ])
print ("Original set           --> ", s)
print ("Num of elements in set --> ", len(s))


---
### **14.2 Opertions on Sets**
---

In [ ]:
a = set(['P', 'Q', 'R', 'S' ])
b = set(["P", "T", "U", "R"])
print ("Original set a      --> ", a)
print ("Original set b      --> ", b)
print ("Union of a and b    --> ", a.union(b))
print ("Intersection of a,b --> ", a.intersection(b))
print ("Difference a - b    --> ", a - b)
print ("Difference a - b    --> ", a.difference(b))
print ("Difference b - a    --> ", b - a)
print ("Difference b - a    --> ", b.difference(a))
print ("Symetric Diff a - b --> ", a.symmetric_difference(b))
print ("Symetric Diff b - a --> ", b.symmetric_difference(a))


---
### **14.3 Add, delete, pop element from set**
---

In [ ]:
a = set(['P', 'Q', 'R', 'S' ])
print ("Original set a       --> ", a)
a.add("T")
print ("Set After Adding (T) --> ", a)
a.add("T")
print ("Set After Adding (T) --> ", a)
a.remove("T")
print ("Set After Deleting(T)--> ", a)
a.pop()
print ("Set After pop        --> ", a)
a.pop()
print ("Set After pop        --> ", a)


---
# **15 Command Line Argument**

**Learning:** How to Take input from command line and process it

**Note:** Run the program at cmd line

---
### **15.1 Add two numbers given at cmd line**
**Note:** To run the program at cmd line
*     python Program.py 10 20

In [ ]:
import sys
print (sys.argv)
a = int(sys.argv[1]) 	# First Number
b = int(sys.argv[2])	# Second Number
c = a + b
print (a, " + ", b, " --> ", c)


---
### **15.2 Concatinate two strings given at cmd line**
**Note:** To run the program at cmd line
*   python Program.py FirstString SecondString

---

In [ ]:
import sys
print (sys.argv)
s = sys.argv[1] + " " + sys.argv[2]
print (sys.argv[1], " + ", sys.argv[2], " --> ", s)


---
### **15.3 Add all the numbers given at cmd line**
**Note:** To run the program at cmd line
*   python Program.py
*   python Program.py 10
*   python Program.py 10 20 30 40

---

In [ ]:
import sys
print (sys.argv)
sum=0
for s in sys.argv[1:]:
	sum += int(s)

print ("Sum is --> ", sum)


---
# **16 File Handling**

**Learning:** How to open the file, read the file and write in the file

---
### **16.1 Writing 1 to 10 in file**

In [ ]:
fp=open('result.txt','w')	# Open the file in writing mode
for  i in range(1,21):
	fp.write(str(i) + "\n")	# Writing to the file line by line
fp.close()

print ("Writing done !! \nOpen result.txt to view the content")


---
### **16.2 Read a file and print its content**
---


In [ ]:
fp=open('result.txt')		# Open the file in reading mode
for line in fp: 		    # print line by line
	print (line.strip())
fp.close()


---
### **16.3 Read from one file, Convert it to upper case and write to other file**
---

In [ ]:
Readfp=open('result.txt')		# Open the file in reading mode
Writefp=open('abc.txt','w')	# Open the file in writing mode
for line in Readfp:
	Writefp.write(line.upper())

Writefp.close()
Readfp.close()

print ("Writing done !! \nOpen result.txt to view the content")
